### PR TITLE
Add instrumentation to freeform question rendering

### PR DESCRIPTION
**Pull request by nwalters512**
This PR adds some instrumentation to the `renderPanel` function used for freeform questions. I'm doing this primarily in preparation for #1626, where we'll want to have detailed visibility into how that's performing, both in terms of errors and timing.

### CODE REVIEW

File: question-servers/freeform.js [hunk 1]
```diff
@@ -1173,13 +1190,47 @@ module.exports = {
...
+    panel,
+    pc,
+    submission,
+    variant,
+    question,
+    course,
+    locals,
+    context
+  ) {
+    return instrumented(`freeform.renderPanel:${panel}`, async (span) => {
```
Comment (nwalters512) [inline question-servers/freeform.js:1212]: Currently, I only have instrumentation for renders, not any of the other operations we have (`generate`, `parse`, `grade`, etc.). Adding instrumentation has proven to be easiest with Promise-based functions, and I didn't want this PR to explode into a full-blown refactor of this code. I intend to do that in a separate PR and then come back to add instrumentation to `processQuestion`, which should give us much better visibility than just instrumenting renders.

File: packages/opentelemetry/src/index.ts [hunk 2]
```diff
@@ -220,8 +244,33 @@ export async function init(config: OpenTelemetryConfig) {
 export async function shutdown(): Promise<void> {
   if (tracerProvider) {
     await tracerProvider.shutdown();
+    tracerProvider = null;
   }
 }
 
+export async function instrumented<T>(
```
Comment (nwalters512) [inline packages/opentelemetry/src/index.ts:251]: This is the interesting bit, the other changes to this file were mostly to make it possible to test functionality here. I'm honestly surprised that this functionality isn't part of the core OTel API!

File: packages/opentelemetry/src/index.ts [hunk 3]
```diff
@@ -202,11 +227,10 @@ export async function init(config: OpenTelemetryConfig) {
     );
   }
 
-  const tracerProvider = new NodeTracerProvider({
+  tracerProvider = new NodeTracerProvider({
```
Comment (nwalters512) [inline packages/opentelemetry/src/index.ts:230]: I think this was a bug 🙃 without this, `tracerProvider` would always be falsy in `shutdown()`.

File: packages/opentelemetry/README.md [hunk 4]
```diff
@@ -23,13 +23,28 @@ await init({
...
 
-To manually instrument code, you can use the `trace` export:
+To easily instrument individual pieces of functionality, you can use the `instrumented()` helper function:
 
 ```ts
-import { trace } from '@prairielearn/opentelemetry';
+import { instrumented } from '@prairielearn/opentelemetry';
+
+async function doThing() {
+  return instrumented('span.name', async (span) => {
```
Comment (mwest1066) [inline packages/opentelemetry/README.md:32]: Nice!

File: packages/opentelemetry/src/index.ts (hunk 3 above)
Comment (mwest1066) [reply packages/opentelemetry/src/index.ts:230]: Yes! Makes sense.

File: packages/opentelemetry/src/index.ts (hunk 2 above)
Comment (mwest1066) [reply packages/opentelemetry/src/index.ts:251]: I really like this kind of wrapper. I've been enjoying `runInTransaction()` recently which is similar.

File: question-servers/freeform.js [hunk 5]
```diff
@@ -1173,13 +1190,47 @@ module.exports = {
...
+      const result = await promisify(module.exports.renderPanel)(
+        panel,
+        pc,
+        variant,
+        submission,
+        course,
+        locals,
+        context
+      );
+      span.setAttribute('cache.status', result.cacheHit ? 'hit' : 'miss');
```
Comment (mwest1066) [inline question-servers/freeform.js:1229]: Cool!

**Review by mwest1066 — APPROVED** [review]

File: packages/opentelemetry/src/index.ts (hunk 2 above)
Comment (nwalters512) [reply packages/opentelemetry/src/index.ts:251]: Yes! Me too. I with there was language-level support for this like Python's `with ...` syntax, but the function form isn't so bad either.

File: question-servers/freeform.js (hunk 5 above)
Comment (nwalters512) [reply question-servers/freeform.js:1229]: Yes! Once we prove this out in prod, I'd like to rip out the cache tracking that we currently have in the `page_view_logs` table.

File: packages/opentelemetry/src/index.ts (hunk 2 above)
Comment (mwest1066) [reply packages/opentelemetry/src/index.ts:251]: That would be so nice!

File: question-servers/freeform.js (hunk 5 above)
Comment (mwest1066) [reply question-servers/freeform.js:1229]: Totally agreed.

**Pull request merged**